crypto/evp/signature.c

```diff
@@ -221,48 +274,110 @@ static void *evp_signature_from_algorithm(int name_id,
                 = OSSL_FUNC_signature_settable_ctx_md_params(fns);
             smdparamfncnt++;
             break;
+        case OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPES:
+            if (signature->query_key_types != NULL)
+                break;
+            signature->query_key_types
+                = OSSL_FUNC_signature_query_key_types(fns);
+            break;
         }
     }
-    if (ctxfncnt != 2
-        || (signfncnt == 0
-            && verifyfncnt == 0
-            && verifyrecfncnt == 0
-            && digsignfncnt == 0
-            && digverifyfncnt == 0
+    /*
+     * In order to be a consistent set of functions we must have at least
+     * a set of context functions (newctx and freectx) as well as a set of
+     * "signature" functions.  Because there's an overlap between some sets
+     * of functions, counters don't always cut it, we must test known
+     * combinations.
+     * We start by assuming the implementation is valid, and then look for
+     * reasons it's not.
+     */
+    valid = 1;
+    /* Start with the ones where counters say enough */
+    if (ctxfncnt != 2)
+        /* newctx or freectx missing */
+        valid = 0;
+    if (valid
+        && ((gparamfncnt != 0 && gparamfncnt != 2)
+            || (sparamfncnt != 0 && sparamfncnt != 2)
+            || (gmdparamfncnt != 0 && gmdparamfncnt != 2)
+            || (smdparamfncnt != 0 && smdparamfncnt != 2)))
+        /*
+         * Params functions are optional, but if defined, they must
+         * be pairwise complete sets, i.e. a getter must have an
+         * associated gettable, etc
+         */
+        valid = 0;
+    if (valid && (initfncnt == 0 || fncnt < initfncnt))
+        /*
+         * No init functions, or fewer performance functions than init
+         * functions
+         */
+        valid = 0;
+
+    /* Now we check for function combinations */
+    if (valid
+        && ((signature->sign_init != NULL
+             && signature->sign == NULL)
+            || (signature->sign_message_init != NULL
+                && signature->sign == NULL
```

Comment:
not sure this is a prereq.. And if its not then counters are sufficient i.e message_init, message_update,message_final use 1 counter

Reply:
I found these more directly readable than the counters, especially since there are overlapping uses.  Point is that if there's a messsage init, there should be *some* associated functionality, a oneshot or update+final, or both.